src/store.rs

```diff
@@ -1053,323 +1179,231 @@ impl StoredObject {
-    fn from(err: TransactionError<Result<(), Failed>>) -> Self {
-        UpdateError(match err {
-            TransactionError::Abort(abort)
-                => ConflictableTransactionError::Abort(abort),
-            TransactionError::Storage(err)
-                => ConflictableTransactionError::Storage(err)
-        })
-    }
-}
-
-
-//------------ Helper Functions ----------------------------------------------
-
-/// Takes an encoded `Time` value from the beginning of a bytes value.
+/// Cleans up a directory tree.
 ///
-/// Upon success, the decoded time will be returned and `bytes` will have
-/// been modified to start after the encoded time.
-fn take_time(bytes: &mut Bytes) -> Result<Time, ObjectError> {
-    if bytes.len() < mem::size_of::<i64>() {
-        return Err(ObjectError)
-    }
-
-    let int_bytes = bytes.split_to(mem::size_of::<i64>());
-    let int = i64::from_be_bytes(int_bytes.as_ref().try_into().unwrap());
-    Ok(Utc.timestamp(int, 0).into())
-}
-
+/// If the closure returns `Ok(false)` for a file with the given path, the
+/// file will be deleted. If all files in a directory are deleted, that
+/// directory is deleted.
+fn cleanup_dir_tree(
+    base: &Path,
+    mut keep: impl FnMut(&Path) -> Result<bool, io::Error>
+) -> Result<(), Failed> {
+    fn recurse(
+        base: &Path,
+        op: &mut impl FnMut(&Path) -> Result<bool, io::Error>
+    ) -> Result<bool, Failed> {
+        let dir = match fs::read_dir(base) {
+            Ok(dir) => dir,
+            Err(err) if err.kind() == io::ErrorKind::NotFound => {
+                return Ok(false)
+            }
+            Err(err) => {
+                error!(
+                    "Failed to read directory {}: {}",
+                    base.display(), err
+                );
+                return Err(Failed)
+            }
+        };
 
-/// Takes an encoded sequence length from the beginning of a bytes value.
-///
-/// Upon success, the decoded length will be returned and `bytes` will have
-/// been modified to start after the encoded length.
-///
-/// All lengths are encoded as `u32` in network byte order, even if we
-/// return `usize` for convenience.
-fn take_encoded_len(bytes: &mut Bytes) -> Result<usize, ObjectError> {
-    if bytes.len() < mem::size_of::<u32>() {
-        return Err(ObjectError)
+        let mut keep = false;
+        for entry in dir {
+            let entry = match entry {
+                Ok(entry) => entry,
+                Err(err) => {
+                    error!(
+                        "Failed to read directory {}: {}",
+                        base.display(), err
+                    );
+                    return Err(Failed)
+                }
+            };
+            let ftype = match entry.file_type() {
+                Ok(ftype) => ftype,
+                Err(err) => {
+                    error!(
+                        "Failed to read directory {}: {}",
```

Comment:
not a directory per se?
I guess the main thing of these last comments is: more specific/detailed error messages might be useful when the user actually needs them for troubleshooting what's wrong with their filesystem